### PR TITLE
Fix crane digest for v1 and platform option

### LINKED ISSUE #906: crane digest does not work with specific platform on v1 schema
crane digest does not work with the platform option on v1 schema images. It returns the following error:
```
unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws", see https://github.com/google/go-containerregistry/issues/377
```

Code to reproduce:
```
crane.Digest("nginx:1.10.0", crane.WithPlatform(&v1.Platform{OS: "linux", Architecture: "amd64"}))
```

Context:
I use crane to sync images from external registries to a local registry. For that I call `crane.Digest(src)` and `crane.Digest(dst)` and if the values differ I do a `crane.Copy` and after the copy trigger some pipelines. I used to use `Copy` and `Digest` without the platform option which worked for v1 schema but this way it did not work for indexes which contain images with non-distributable layers.

### PR DESCRIPTION
**Pull request by dvob**
This allows to use crane.Digest on images with a v1 schema.

Fixes #906

### CODE REVIEW

**Comment by dvob** [comment]
@googlebot I signed it!

File: pkg/crane/digest.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ func Digest(ref string, opt ...Option) (string, error) {
 	o := makeOptions(opt...)
 	if o.platform != nil {
 		desc, err := getManifest(ref, opt...)
+		if !desc.MediaType.IsIndex() {
```
Comment (ImJasonH) [inline pkg/crane/digest.go:22]: It looks like desc will be `nil` if there are errors fetching the manifest (e.g., HTTP 404, 403), which will result in a nil panic at this line. Should this check happen after the err check?

**Review by ImJasonH — DISMISSED** [review]

File: pkg/crane/digest.go (hunk 1 above)
Comment (dvob) [reply pkg/crane/digest.go]: Yes this should be after the err check. I moved it.

**Comment by dvob** [comment]
How often do you release new versions (tags)?

**Comment by ImJasonH** [comment]
> How often do you release new versions (tags)?

We tend to release as-needed whenever there are significant changes to make available. 

cc @jonjohnsonjr to speak on whether/how we want `crane` to support v1 manifests. This change by itself seems innocuous to me at least 🤷

**Comment by jonjohnsonjr** [comment]
> How often do you release new versions (tags)?

Yeah, roughly as needed. Usually when I need to re-import this package into google or some large downstream dependency requests a fix be included in a release. We could have more discipline about this (cc @hasheddan).

> This change by itself seems innocuous to me at least 🤷

This doesn't cover every base, but it's definitely an improvement, so I'm fine with it.

@dvob consider situations such as this: https://github.com/docker-library/golang/issues/269#issuecomment-729984068

```
$ crane manifest nginx:1.19.0-alpine | jq .
2021/01/08 09:18:20 No matching credentials were found for "index.docker.io/library/nginx", falling back on anonymous
{
  "manifests": [
    {
      "digest": "sha256:ee5a9b68e8d4a4b8b48318ff08ad5489bd1ce52b357bf48c511968a302bc347b",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      },
      "size": 1360
    },
    {
      "digest": "sha256:88fd834bc089b5dff1f27ab0315c3b0bbdfddfca3b2c1aeb9798944d179e441b",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "arm",
        "os": "linux",
        "variant": "v6"
      },
      "size": 1360
    },
    {
      "digest": "sha256:bf8ebb8068d7ff10711c1670398eeaa83283a7987cdc9b9b014560dbba71773c",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "arm",
        "os": "linux",
        "variant": "v7"
      },
      "size": 1360
    },
    {
      "digest": "sha256:eff196a3849ad6541fd3afe676113896be214753740e567575bb562986bd2cd4",
      "mediaType": "application/vnd.docker.distribution.manifest.v1+json",
      "platform": {
        "architecture": "arm64",
        "os": "linux",
        "variant": "v8"
      },
      "size": 13297
    },
    {
      "digest": "sha256:2fa12030ffb0224e0b2e17bc4b1f1479e191e2fd65869fc60d09a8efa5b6d879",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "386",
        "os": "linux"
      },
      "size": 1360
    },
    {
      "digest": "sha256:9012a1962d3201492ea815d0b395af623cf78e4b2187f9069d5cee0c0d1345c1",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      },
      "size": 1360
    },
    {
      "digest": "sha256:d7b9368659ff1be2bbb4267c9d45e7aa201cb282ac9bc86e09bf390087df68b6",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "s390x",
        "os": "linux"
      },
      "size": 1360
    }
  ],
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "schemaVersion": 2
}
```

The linux/arm64/v8 image is a schema 1 image, so `desc.Image()` will fail, even though we could just pluck the digest from that descriptor. I think this is such a rare case that it's probably not worth fixing (and is currently broken anyway), but possibly add a TODO?

**Comment by dvob** [comment]
Or do you see an easy way to fix it for these cases as well?

**Comment by jonjohnsonjr** [comment]
> Or do you see an easy way to fix it for these cases as well?

We'd need to move this logic somewhere else:

https://github.com/google/go-containerregistry/blob/f97c411e4aa61a02a2fefd761529d14106687a83/pkg/v1/remote/index.go#L193-L237

(probably as a `match.Matcher`)

And use that inside `crane.Digest` to match the right platform. Like here:

https://github.com/google/go-containerregistry/blob/f97c411e4aa61a02a2fefd761529d14106687a83/pkg/v1/remote/index.go#L143-L157

The TODO is fine for now, I think. There is some discussion around platform matching in https://github.com/google/go-containerregistry/issues/892 that we can tack this work onto.

**Review by jonjohnsonjr — APPROVED** [review]

**Pull request merged**